Sites Management Page: Display a nag when site's plan is expired

kozer opened the pull request:
#### Proposed Changes

Display notice on the Sites Management Page when a site's plan is expired.

If the user is the site owner, there will be a "Renew" button, so that he will be able to renew the
site:

![site_is_owner_b](https://user-images.githubusercontent.com/497103/195867326-e5e9e344-f98c-4d3a-9085-0872d91726cd.png)


If not, then only the expiration message should be shown:

![site_is_not_owner_b](https://user-images.githubusercontent.com/497103/195867374-7124d561-1d9c-4f59-b647-9982289ec7c8.png)


#### Testing Instructions

##### Site onwer test
- Create a new Business site.
- Take the site Atomic.
- Set the site's plan expiry to a day before the plan was added to the site.
- Locally, visit `/sites`.
- Confirm that Renew button and your plan followed by the word "Expired" exits.

##### Not site owner test
- Create a new Business site.
- Take the site Atomic.
- Set the site's plan expiry to a day before the plan was added to the site.
- Invite a user to your site.
- Login with this user.
- Locally, visit `/sites`.
- Confirm that your plan followed by the word "Expired" exits.
- Confirm that Renew button doesn't appear on the page

##### Track events test

- Create a new Business site.
- Take the site Atomic.
- Set the site's plan expiry to a day before the plan was added to the site.
- Locally, visit `/sites`.
- Open browser developer tools Console
- Copy and paste this into the Console:  `localStorage.setItem( 'debug', 'calypso:analytics*' );`
- Enable all items in the filters by clicking on “All levels” and ensuring all items are checked.
- Refresh the page.
- On filter text write: `rec renew_nag`
- Make sure that `Record event 'calypso_sites_dashboard_plan_renew_nag_inview'` appears in the console.
- Click on the "Renew button"
- Make sure that `Record event 'calypso_sites_dashboard_plan_renew_nag_click'` appears in the
  console.

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [X] Have you checked for TypeScript, React, or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to 1017-gh-Automattic/dotcom-forge

kozer committed:
feature: Add `isExpiredAtomic` helper function

kozer committed:
feature: Add site renew component to display expiration info

kozer committed:
update: Add new site renew component to table row view

kozer committed:
update: Show renew button only if the user is the site owner

kozer committed:
update: Add tracking on site renew component

- Add track on click the renew button
- Add track on view the component

danielbachhuber commented on client/sites-dashboard/components/sites-site-renew.tsx:
```diff
@@ -0,0 +1,64 @@
+
+	useEffect( () => {
+		if ( inViewOnce ) {
+			recordTracksEvent( SITE_RENEW_NAG_IN_VIEW );
```
It'd be great to put a `site_owner` boolean property on this Tracks event to distinguish between whether the nag is shown to a site owner or not.

danielbachhuber commented on client/sites-dashboard/components/sites-site-renew.tsx:
```diff
@@ -0,0 +1,64 @@
+		<SiteRenewContainer ref={ ref }>
+			<SiteRenewNotice>
+				<Gridicon icon="notice" />
+				{ `${ site.plan?.product_name_short } - Expired` }
```
This string probably should be localized?

danielbachhuber commented on client/sites-dashboard/components/sites-table-row.tsx:
```diff
@@ -130,7 +137,11 @@ export default memo( function SitesTableRow( { site }: SiteTableRowProps ) {
 						</SitePlanIcon>
 					) }
-					{ site.plan?.product_name_short }
+					{ isExpiredAtomic( site ) ? (
```
Should we show this for all expired sites, not just Atomic?

kozer replied:
Not sure about that. This is what I understood from the description of the issue.
Maybe @danielbachhuber can help here?

danielbachhuber replied:
@kozer Hi, original commenter here 😁

I think we can show this nag for all expired sites, not just expired Atomic sites. Given we'd like users to renew, I think it makes sense to show it to more applicable users.

kozer replied:
Oooopsie @danielbachhuber  :D . So the first "is done" doesn't apply? 
It states:
"When an Atomic site's plan is expired, but the site isn't reverted, we display a notice on the Sites Management Page."

danielbachhuber replied:
> Oooopsie @danielbachhuber :D . So the first "is done" doesn't apply? It states: "When an Atomic site's plan is expired, but the site isn't reverted, we display a notice on the Sites Management Page."

It broadens the scope a bit: _When an ~Atomic~ site's plan is expired, ~but the site isn't reverted,~ we display a notice on the Sites Management Page._

kozer replied:
Great! thanks for the explanation! Will make the change!

danielbachhuber replied:
@kozer Yep! Sorry I wasn't clear enough in the first place.

danielbachhuber commented on client/sites-dashboard/components/sites-site-renew.tsx:
```diff
@@ -0,0 +1,64 @@
+import { recordTracksEvent } from '@automattic/calypso-analytics';
```
Can we rename this file to `sites-site-renew-nag`, for consistency with `sites-site-launch-nag`?

danielbachhuber reviewed (commented):
Looks like a great start! I left a few comments inline.

Also, design-wise, what do you think about making it more consistent with the 'Launch checklist' donut?

<img width="597" alt="image" src="https://user-images.githubusercontent.com/36432/195373358-71a73dbf-f736-46fd-bec1-7f92f4ebe344.png">

Non-site owners could simply see 'Plan expired' instead of 'Renew plan'.

zaguiini commented on client/sites-dashboard/components/sites-site-renew.tsx:
```diff
@@ -0,0 +1,64 @@
+	display: flex;
+	align-items: center;
+	color: #ea303f;
+	margin-top: -6px;
```
Do we need this negative top margin? It looked centralized when I tested it:

![image](https://user-images.githubusercontent.com/26530524/195375000-08d1238f-b85f-4274-ae3f-7f763f8e687b.png)

Also, how about adding a 4px margin to the right of the icon? It'll give more room for the elements to breathe:

![image](https://user-images.githubusercontent.com/26530524/195375142-b3f09703-b89c-45f2-977d-2885027b922a.png)

kozer replied:
Changed it in order to target all expired plans!
It's a good idea the adding some margin there. Will do.

danielbachhuber replied:
@kozer Can we [improve the design a bit](https://github.com/Automattic/wp-calypso/pull/68965#pullrequestreview-1139208581)?

kozer replied:
This version confuses me a bit.
Maybe let the icon be next to the plan name and change the wording as you proposed?
Also, I think the color was ok at the plan name. It draws the attention onto what happened.
What do you think @danielbachhuber ?

danielbachhuber replied:
@kozer Personally, I prefer the visual consistency of my suggested alternative. I don't know that I have a strong opinion right now, though. I've asked for feedback in Slack: p1665669234361159-slack-C041RHH38NQ

vykes-mac replied:
posting the suggested design from the slack thread here.
![image](https://user-images.githubusercontent.com/47489215/195660257-2934897a-4c29-458d-a612-f60ada2d4ff7.png)

zaguiini reviewed (commented):
It works fine 👍  Why are we only accounting for Atomic sites here, though? You can have a simple site with a paid plan.

kozer committed:
update: Changes based on review

- Rename `sites-site-renew` to `sites-site-renew-nag`
- Localized `Expired` string
- Add `is_site_owner` property to track in view event
- Add 4px margin to the notice icon.
- Show the notice in all expired sites. Not only on atomics.

zaguiini commented on client/sites-dashboard/components/sites-site-renew-nag.tsx:
```diff
@@ -0,0 +1,69 @@
+	const { __ } = useI18n();
+	const { ref, inView: inViewOnce } = useInView( { triggerOnce: true } );
+
+	const isSiteOwner = site.plan?.user_is_owner || false;
```
Usually, null coalescing (`??`) instead of "or short-circuiting" (`||`) is preferred because `||` catches any falsy value instead of only `null` and `undefined`. `user_is_owner` is a boolean value so we want to use it if `plan` is defined, so:

```suggestion
	const isSiteOwner = site.plan?.user_is_owner ?? false;
```

Let's use null coalescing instead.

mpkelly replied:
Personally, I think `||` is better here because it will always convert invalid falsy values, e.g. `0`, to `boolean` while `??` will keep them. But I think your other comment about changing the props so `plan` is known to be `non-null` is the best solution. 👍

kozer replied:
@mpkelly and @zaguiini  this is the reason I added `||` instead of `??`. Although this will get removed as I'll go with checking that the plan is `non-null` in the first place!
Thank you both!

zaguiini commented on client/sites-dashboard/components/sites-site-renew-nag.tsx:
```diff
@@ -0,0 +1,69 @@
+		<SiteRenewContainer ref={ ref }>
+			<SiteRenewNotice>
+				<Gridicon icon="notice" />
+				{ `${ site.plan?.product_name_short } - ${ expiredText }` }
```
This string is also translatable. We need to use the `sprintf` function to account for placeholders on translatable strings. So we should change this to:

```ts
{ sprintf( __( '%s - Expired' ), site.plan?.product_name_short ) }
```

This way the placeholders/variables are taken into account when translating the string to RTL languages for example.

vykes-mac replied:
I think this is a good candidate to add as part of a checklist 😄 . This is not always evident when persons are just starting out but a checklist item would help re-inforce this requirement.

zaguiini commented on client/sites-dashboard/components/sites-site-renew-nag.tsx:
```diff
@@ -0,0 +1,69 @@
+	}
+`;
+
+export const SiteRenewNag = ( { site }: SiteRenewProps ) => {
```
One thing we could do to avoid those ugly null-coalescing and optional chaining operators is to pass the `plan` as a required prop instead of the whole `site` object, this way making the surface area/scope of the component smaller since we're not really using the other props.

vykes-mac replied:
I'm in agreement with this, by reducing the usage of null-coalescing we also reduce the surface area for potential bugs. Make it a required prop and allow the parent component to render this component only when a plan is available.

mpkelly replied:
I agree with this change - make the API as simple as possible so it's easier to test and reuse.

kozer replied:
Started a discussion on slack about this! Thanks @zaguiini !

kozer replied:
There is an issue with simply redirecting the user to the checkout, it may be the case that the user has en empty card.
I started a discussion p1665755433194219-slack-C7YPUHBB2 for that.

mpkelly commented:
Looking good. I have tested this on Chrome and Firefox. Here's what I found:

As per the Slack thread, the icon size should be reduced. Sometimes we will output a Jetpack icon which has a size of `16px`. It should be the same I think. But this `notice` icon is odd and has some padding so I found making it `20px` is about right. 

![image](https://user-images.githubusercontent.com/6851384/195769779-26c8464b-c2eb-4411-bcd0-e07f9d84c428.png)


On smaller screens (<1100px) the columns start to overlap. I think we should consider hiding one or more columns when space is tight. 

![image](https://user-images.githubusercontent.com/6851384/195769760-18a0a9a8-c479-4bd7-8688-dd56598454ee.png)

kozer committed:
update: Update component to reflect the new UI mocks

kozer committed:
update: Use sprintf to account for placeholders on translatable strings

kozer commented:
> 

I did a small change in order to, at least, wrap the text so that it is presented better! Thanks for catching that!

kozer committed:
update: Change component props

- Pass `plan` and `checkoutUrl` as props instead of site
- Change component name to better reflect what it does based on the new
  props.

kozer committed:
fix: Add the `plan_slug` to the checkout url in order to autocreate the
card.

danielbachhuber commented on client/sites-dashboard/components/sites-table-row.tsx:
```diff
@@ -130,7 +131,14 @@ export default memo( function SitesTableRow( { site }: SiteTableRowProps ) {
+					{ site.plan?.expired ? (
+						<PlanRenewNag
+							plan={ site.plan }
+							checkoutUrl={ `/checkout/${ site.slug }/${ site.plan?.product_slug }` }
```
Does this URL work for Jetpack plans too?

danielbachhuber commented on client/sites-dashboard/components/sites-plan-renew-nag.tsx:
```diff
@@ -0,0 +1,81 @@
+						href={ checkoutUrl }
+						title={ renewText }
+					>
+						{ renewText }
```
What do you think about matching the 'Renew' font size and line height to 'Launch checklist'? It might also be better as 'Renew plan', and to have both lines of text match up (couldn't figure that out for my mockup)

## Before

<img width="554" alt="image" src="https://user-images.githubusercontent.com/36432/195891916-b8503a2b-d8cc-42d4-8860-897eae330c7e.png">

## After

<img width="562" alt="image" src="https://user-images.githubusercontent.com/36432/195892252-e391f494-10e6-4962-a321-aaa84191d3cc.png">




### After

danielbachhuber commented on client/sites-dashboard/components/sites-plan-renew-nag.tsx:
```diff
@@ -0,0 +1,81 @@
+} );
+
+const IconContainer = styled.div`
+	color: #ea303f;
```
Are there CSS variables we should use for colors?

kozer replied:
I couldn't find any for this color.

danielbachhuber reviewed (commented):
Looking much better, @kozer ! Left some comments inline.

I appreciate you updating the screenshots in your description 😊

kozer committed:
update: Update plan renew nag component link font size, and text to reflect pr discussions

kozer commented:
For jetpack sites, the checkout seems to work as expected (thanks @danielbachhuber for mention to test it), although I found a strange issue with the icon

![jetpack_strange_icon_issue](https://user-images.githubusercontent.com/497103/196229034-b8747923-12d7-4bb6-a483-01faebb50206.png)

Although the jetpack icon is aligned with the middle of the text, it looks ugly next to the notice icon.
A way I found to make it better is that in the case of a nonatomic jetpack site for an expired plan, to render the icon inside the plan
nag component, but still, if I leave it to 16 px looks a bit off. 
I tried adding a margin of `1px` and `2px` as well, but still, it's not perfectly aligned.
What do you think is the best way to tackle this?

cc: @danielbachhuber , @mpkelly

danielbachhuber commented:
@kozer For now, let's simply hide the Jetpack icon when the plan is expired.

kozer committed:
update: Remove jetpack icon from site table when plan is expired

danielbachhuber commented on client/sites-dashboard/components/sites-plan-renew-nag.tsx:
```diff
@@ -0,0 +1,82 @@
+	justify-items: flex-start;
+	gap: 4px;
+	margin-top: -5px;
+`;
```
Should this CSS be defined as an object here too?

sejas replied:
For context, a couple of months ago, there was a conversation in favor of using CSS objects. 💪.
p1660141381467549-slack-C0347E545HR

danielbachhuber commented on client/sites-dashboard/components/sites-plan-renew-nag.tsx:
```diff
@@ -0,0 +1,82 @@
+
+	useEffect( () => {
+		if ( inViewOnce ) {
+			recordTracksEvent( PLAN_RENEW_NAG_IN_VIEW, { is_site_owner: isSiteOwner } );
```
Love the `is_site_owner` prop 😁

Let's also pass `product_slug: plan.product_slug`.

Not sure if you're already aware, but you'll need to register these two new events in Tracks.

kozer replied:
I wasn't aware! Thanks for letting me know!

danielbachhuber commented on client/sites-dashboard/components/sites-table-row.tsx:
```diff
@@ -130,13 +131,19 @@ export default memo( function SitesTableRow( { site }: SiteTableRowProps ) {
+					{ site.plan?.expired ? (
+						<PlanRenewNag
+							plan={ site.plan }
+							checkoutUrl={ `/checkout/${ site.slug }/${ site.plan?.product_slug }` }
```
```suggestion
							checkoutUrl={ `/checkout/${ site.slug }/${ site.plan.product_slug }` }
```
Shouldn't `site.plan` always exist by this condition?

danielbachhuber commented on client/sites-dashboard/components/sites-plan-renew-nag.tsx:
```diff
@@ -0,0 +1,82 @@
+	flex-direction: column;
+	color: #ea303f;
+	gap: 4px;
+`;
```
These two elements still aren't quite aligned:

<img width="463" alt="image" src="https://user-images.githubusercontent.com/36432/196425345-621f076b-9ef6-4833-8005-44c1e90eec03.png">

Is it possible to have them on the same plane?

Also, on smaller widths, we probably should ellipsis the "<plan> - Expired" for consistency with "Coming Soon":

<img width="779" alt="image" src="https://user-images.githubusercontent.com/36432/196425560-d1e50d12-e1d5-4fdf-bfb1-64985ea734b7.png">

We can expand to wrapped text on larger widths.

danielbachhuber commented on client/sites-dashboard/components/sites-plan-renew-nag.tsx:
```diff
@@ -0,0 +1,82 @@
+				{ isSiteOwner && (
+					<PlanRenewLink
+						onClick={ () => {
+							recordTracksEvent( PLAN_RENEW_NAG_ON_CLICK );
```
Let's also pass `{ product_slug: plan.product_slug }` as a property here.

danielbachhuber reviewed (commented):
Pretty close to the finish line! Just a couple more things to wrap up.

sejas commented on client/sites-dashboard/components/sites-table-row.tsx:
```diff
@@ -130,13 +131,19 @@ export default memo( function SitesTableRow( { site }: SiteTableRowProps ) {
+							plan={ site.plan }
+							checkoutUrl={ `/checkout/${ site.slug }/${ site.plan?.product_slug }` }
+						/>
+					) : (
+						<>{ site.plan?.product_name_short }</>
+					) }
```
I suggest removing the unnecessary fragment.

```suggestion
					) : (
						site.plan?.product_name_short
					) }
```

sejas reviewed (commented):
@kozer , Well done !!

I've tested different scenarios, and it's Working well on Pro, Business, and Personal.

The only difference with the testing instructions that I found is on `Not site owner test`.
I've transferred a simple site with the personal plan to a different account, and  I can see the `Renew plan` button.
So, `Confirm that Renew button doesn't appear in the page` is false in my case.
<img width="153" alt="Screenshot 2022-10-19 at 10 42 35" src="https://user-images.githubusercontent.com/779993/196656598-7612e8df-f6a4-4b4e-a355-8e65db3df219.png">

I couldn't [transfer an atomic site](https://wordpress.com/support/transferring-a-site-to-another-wordpress-com-account/#for-plugin-enabled-sites).



I've updated the testing instructions with the actual tracking variable names to match the code, using `plan` instead of `site`:
- `calypso_sites_dashboard_site_renew_nag_inview` -> `calypso_sites_dashboard_plan_renew_nag_inview `
- `calypso_sites_dashboard_site_renew_nag_click` -> `calypso_sites_dashboard_plan_renew_nag_click `

kozer committed:
update: Add ellipsis to expire text instead of wrapping

kozer committed:
update: Change all styled components to objects

kozer committed:
fix: Fix plan nag component align issues

kozer committed:
update: Add product_slug to plan nag track events

kozer committed:
update: Remove unecessary fragment in sites table row component

kozer commented:
> @kozer , Well done !!
> 
> I've tested different scenarios, and it's Working well on Pro, Business, and Personal.
> 
> The only difference with the testing instructions that I found is on `Not site owner test`. I've transferred a simple site with the personal plan to a different account, and I can see the `Renew plan` button. So, `Confirm that Renew button doesn't appear in the page` is false in my case. <img alt="Screenshot 2022-10-19 at 10 42 35" width="153" src="https://user-images.githubusercontent.com/779993/196656598-7612e8df-f6a4-4b4e-a355-8e65db3df219.png">
> 
> I couldn't [transfer an atomic site](https://wordpress.com/support/transferring-a-site-to-another-wordpress-com-account/#for-plugin-enabled-sites).
> 
> I've updated the testing instructions with the actual tracking variable names to match the code, using `plan` instead of `site`:
> 
> * `calypso_sites_dashboard_site_renew_nag_inview` -> `calypso_sites_dashboard_plan_renew_nag_inview `
> * `calypso_sites_dashboard_site_renew_nag_click` -> `calypso_sites_dashboard_plan_renew_nag_click `

In order for that to be true, it means that plan's `user_is_owner` prop hasn't been updated, as I am using that to display the button.
I didn't test that case to be honest, as I don't have a second WordPress account and I thought that this is a straightforward use case.
My bad, and thanks for the finding!
Can you help me with that?

sejas reviewed (approved):
@kozer  Thanks for updating the testing instructions. 🙌 
Looks good to me!! Well done!

kozer committed:
fix: Rely on `site_onwer` prop instead of plan's `is_site_owner`

It seems that plan's `is_site_owner` property is not updated
correctly if we transfer the blog through blog RC.
So we changed that to check the site's `site_owner` prop against the current user id in order to check each site's ownership.

danielbachhuber reviewed (approved):
Looks like it's getting clipped but not ellipsis on smaller widths in Chrome:

<img width="375" alt="image" src="https://user-images.githubusercontent.com/36432/196975776-b8272799-37c6-419f-8ff2-4f484051a51e.png">

Feel free to 🚢 once you get this sorted out 😁

kozer commented:
> Looks like it's getting clipped but not ellipsis on smaller widths in Chrome:
> 
> <img alt="image" width="375" src="https://user-images.githubusercontent.com/36432/196975776-b8272799-37c6-419f-8ff2-4f484051a51e.png">
> 
> Feel free to ship once you get this sorted out grin

This is due to the fact that I put the ellipsis only on the `- Expired` part.
Not around the plan name too.
Do we want to add an ellipsis on the plan name as well?

danielbachhuber commented:
> This is due to the fact that I put the ellipsis only on the `- Expired` part.
> Not around the plan name too.
> Do we want to add an ellipsis on the plan name as well?

Yep! Sorry for the confusion on my original comment.

kozer committed:
Merge branch 'trunk' into feature/sites-management-plan-expire-notice

kozer committed:
update: Add ellipsis around all words in expire text

kozer commented:
> as

I was sure that this was gonna be the answer so I already done it! 😁 If you think everything is good, I'll ship it tomorrow!

Thanks!

danielbachhuber commented:
@kozer Looks good! Excited to see it live 😊

kozer committed:
Merge branch 'trunk' into feature/sites-management-plan-expire-notice

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7587658).
